### PR TITLE
Turn on/off progress bar cancellation for generator and solar panel

### PR DESCRIPTION
**Pull request by y0014984**
At the moment if the ACE3 progress bar for turning on/off generator or solar panel is cancelled, the setup is in an undefined state because some code like setting the on/off state is set although the action was cancelled.

By merging this pull request this issue is fixed. Now It's possible to cancel the progress bars. It's not a very elegant solution because we are switching from unscheduled to scheduled execution and therefore the handling of the return value is a problem. But I think the workaround is ok.

Another solution would be getting rid the progress bar at all. I don't think that the solar panel needs a progress bar. It could be turned on/off immediately. Also the generator doesn't need a interruptable stop/stop procedure. I think it's more like pressing a button. The start and stop sounds are only for immersion.

### CODE REVIEW

File: addons/power/functions/fnc_turnOffGeneratorAction.sqf [hunk 1]
```diff
@@ -8,35 +8,67 @@
...
-};
+			[_entity] call _turnOffGenFunc;
 
-[_entity, true, [0, 1, 0], 0] call ace_dragging_fnc_setDraggable;
+			// we need to set power state here because function already returned false
+			// and therefore the turn on wrapper doesn't set the state to turned on
+			_entity setVariable ["AE3_power_powerState", 0, true];
+		},
+		{
+			// following code only runs on progress bar success
```
Comment (GermanHydrogen) [inline addons/power/functions/fnc_turnOffGeneratorAction.sqf:52]: do you mean "on progress bar failure"?

File: addons/power/functions/fnc_turnOffSolarAction.sqf [hunk 2]
```diff
@@ -8,30 +8,61 @@
...
+			[_entity] call _turnOffSolPanFunc;
 
+			// we need to set power state here because function already returned false
+			// and therefore the turn on wrapper doesn't set the state to turned on
+			_entity setVariable ["AE3_power_powerState", 0, true];
 		},
-		{},
-		(localize "STR_AE3_Power_Interaction_TurnOff")
+		{
+			// following code only runs on progress bar success
```
Comment (GermanHydrogen) [inline addons/power/functions/fnc_turnOffSolarAction.sqf:52]: same as above.

File: addons/power/functions/fnc_turnOnSolarAction.sqf [hunk 3]
```diff
@@ -2,28 +2,43 @@
...
+		[_entity, AE3_power_fnc_solarCalculation] remoteExecCall ["AE3_power_fnc_addProviderHandler", 2];
+
+		[_entity, false, [0, 1, 0], 0] call ace_dragging_fnc_setDraggable;
+
+		// we need to set power state here because function already returned false
+		// and therefore the turn on wrapper doesn't set the state to turned on
+		_entity setVariable ["AE3_power_powerState", 1, true];
+	},
+	{
+		// following code only runs on progress bar fail
```
Comment (GermanHydrogen) [inline addons/power/functions/fnc_turnOnSolarAction.sqf:35]: could be empty

File: addons/power/functions/fnc_turnOffGeneratorAction.sqf (hunk 1 above)
Comment (GermanHydrogen) [reply addons/power/functions/fnc_turnOffGeneratorAction.sqf:52]: could be empty

**Review by GermanHydrogen — CHANGES_REQUESTED** [review]

File: addons/power/functions/fnc_turnOffGeneratorAction.sqf (hunk 1 above)
Comment (y0014984) [reply addons/power/functions/fnc_turnOffGeneratorAction.sqf:52]: Yes, you are right. I will fix that.

File: addons/power/functions/fnc_turnOffSolarAction.sqf (hunk 2 above)
Comment (y0014984) [reply addons/power/functions/fnc_turnOffSolarAction.sqf:52]: See above.

File: addons/power/functions/fnc_turnOnSolarAction.sqf (hunk 3 above)
Comment (y0014984) [reply addons/power/functions/fnc_turnOnSolarAction.sqf:35]: Yes, I think I will remove the unnecessary code lines.

**Review by GermanHydrogen — APPROVED** [review]

**Pull request merged**